casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -99,149 +172,236 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-    "C! C!" -> COULD_MATCH_SAME_CONFLICTS -> conflicts(C_, C_, S0),
-    "C! C!" -> COULD_MATCH_SAME_MERGES    -> merges(C0, C1, S0 | S1),
-    // 4!! / !!4 row is similar to !4 / 4! and thus skipped
-    // C!! / !!C row is similar to !C / C! and thus skipped
-    "CX !!X"    -> INCOMING_COULD_MATCH       -> conflicts(R0, C_, Nil),
-    "(!4) !4"   -> VOLATILE_EVENT             -> coveredBy("!4 (!4)"),
-    "(!4) (!C)" -> VOLATILE_EVENT             -> coveredBy("(!4) !C"),
-    "(!4) (4!)" -> VOLATILE_EVENT             -> merges(S0 | F_, S0 | F_, Nil),
-    "(!4) (C!)" -> PERSISTENT_COULD_NOT_MATCH -> merges(S0 | F_, C_ | S0, Nil),
-    "(!4) 4X"   -> VOLATILE_EVENT             -> merges(S0 | F_, F_, Nil),
-    "(!4) CX"   -> VOLATILE_EVENT             -> merges(S0 | F_, C_, Nil),
-    "(!C) (!C)" -> PERSISTENT_COULD_NOT_MATCH -> merges(S0 | C_, S0 | C_, Nil),
-    "(!C) (4!)" -> VOLATILE_EVENT             -> merges(S0 | C_, F_ | S0, Nil),
-    "(!C) (C!)" -> PERSISTENT_COULD_NOT_MATCH -> merges(S0 | C_, C_ | S0, Nil),
-    "(!C) 4!"   -> PERSISTENT_COULD_MATCH     -> conflicts(S0 | C_, F_, S0),
-    "(!C) 4X"   -> PERSISTENT_COULD_NOT_MATCH -> merges(S0 | C_, F_, Nil),
-    "(!C) C!"   -> PERSISTENT_COULD_MATCH     -> conflicts(S0 | C_, C_, S0),
-    "(!C) CX"   -> PERSISTENT_COULD_NOT_MATCH -> merges(S0 | C_, C_, Nil),
-    "(4!) (4!)" -> VOLATILE_EVENT             -> merges(F_ | S0, F_ | S0, Nil),
-    "(4!) (C!)" -> VOLATILE_EVENT             -> merges(F_ | S0, C_ | S0, Nil),
-    "(4!) 4!"   -> VOLATILE_EVENT             -> merges(F1 | S1, F_, S0),
-    "(4!) C!"   -> VOLATILE_EVENT             -> merges(F1 | S1, C_, S0),
-    "(C!) C!"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(C_ | S0, C_, S0),
-    "4! (4!)"   -> VOLATILE_EVENT             -> merges(F_, F1 | S1, S0),
-    "4! (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> conflicts(F_, C_ | S0, S0),
-    "4X (4!)"   -> VOLATILE_EVENT             -> merges(F_, F_ | S0, Nil),
-    "4X (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(F_, C_ | S0, Nil),
-    "C! (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(C_, C1 | S1, S0),
-    "CX (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> merges(C_, C_ | S0, Nil),
-    "(!4) 4!"   -> VOLATILE_EVENT             -> merges(S1 | F1, F_, S0),
-    "(!4) !C"   -> VOLATILE_EVENT             -> merges(S1 | F1, S0, C0),
-    "(!4) C!"   -> VOLATILE_EVENT             -> merges(S0 | F0, C_, S1),
-    "(4!) CX"   -> VOLATILE_EVENT             -> merges(F_ | S0, C_, Nil),
-    "(C!) (C!)" -> PERSISTENT_COULD_NOT_MATCH -> merges(C_ | S0, C_ | S0, Nil)
+    "!X !X"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S0, Nil),
+    "!X !4"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, F1),
+    "!X (!4)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S0, F_)(Nil),
+    "!X !C"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, C1),
```

Comment:
We can skip (get rid of) the `coveredBy` cases, or provide actual examples for them too.

Reply:
What you propose isn't going to work because a `Map` cannot have duplicate keys and so the last entry for `!X !X` is going to win and all previous values are going to be ignored. Originally the description was used to disambiguate between the keys/tags/names. Another way to mitigate this would be to get rid of the `Map` and store tuples in a `List`. The second element of the tuple would be a sequence of cases (of size one for most). WDYT?

Reply:
I am not in favor of introducing a single trait for all test cases. Implementing what I proposed in the comment above feels enough. I agree that it makes sense to split the `CouldMatch*` case into two.

Reply:
Good point! Using a map will cause 'duplicate key silent errors' without means of validating it. I'm against nested lists though. 

We should use a (flat) list, and enforce the names to be unique / make them unique by combining with the list index.

Reply:
> I am not in favor of introducing a single trait for all test cases.

It's more about the duplication of
`!X !4 -> $FOO.desc -> $FOO.{merges/fails}`
instead of having a common interface. The latter is just a means.

I think you'll find the trait needed to obtain the description for the assertion message. Notice also: once you change the `merges/conflicts` to `assertion` (which is needed, as the assertion should be same for all examples of a given case), the common interface is already there. Adding a trait atop is a matter of a couple `extends`.

Reply:
This has a downside of unnecessarily complicating the test which checks whether all cases from the spreadsheet are covered. Hence I would rather go with a list. Unless there is a better argument for what you are proposing

Reply:
I'm fine if we fuse `preconditions` and `assertion` into just `assertion`

Reply:
> This has a downside of unnecessarily complicating the test which checks whether all cases from the spreadsheet are covered.

OTOH having the a list for each case instead of just a single value massively complicates the main test, and puts a lot of strain on the compiler - just to have 4-ish names have two tests instead of one.

The complication you mention is simply:
add the index before doing any set operations (e.g. all cases are one of those expected),
drop the index before doing any exhaustivity checks (e.g. each expected case has at least one representant)

Reply:
Frankly, you'd have coded it up faster than the time we've spent discussing this.

Reply:
> massively complicates the main test, and puts a lot of strain on the compiler 

This is greately exaggerated

> Frankly, you'd have coded it up faster than the time we've spent discussing this.

The comment above could be spared, it doesn't bring anything. We're talking about sensibility of the approaches

Why are you suggesting combining with an index at all? Entries in a `List` do not have to be unique

Reply:
> This is greately exaggerated

Maybe w.r.t. to the compiler part. But usage-wise, reading and editing
`"4! 4!" -> List(CouldMatchSameMerges(X, Y, Z))`
on each line, is going to be a pain. It also makes the actual validations and the main test harder, b/c now you're handling a list of lists.

You're right that combining with an index might not even be needed. I suggested it as a means of regaining uniqueness, if it's needed in validations (e.g. we build a set somewhere). Most likely it's not needed though.

> The comment above could be spared, it doesn't bring anything. We're talking about sensibility of the approaches

I though really believe that the raised problem of `unnecessarily complicating the (spreadsheet coverage) test` would've been cheaper to handle in that test without this discussion. So while I'm all for considering sensibility of approaches, I hate to discuss such minutia.
I don't know, maybe I'm overlooking sth big in there?

Reply:
```
Maybe w.r.t. to the compiler part. But usage-wise, reading and editing
"4! 4!" -> List(CouldMatchSameMerges(X, Y, Z))
on each line, is going to be a pain. It also makes the actual validations and the main test harder, b/c now you're handling a list of lists.
```

Maybe the reason I don't find this harder is that I'm used to handling nested lists and flattening doesn't give me headaches. I find grouping related cases under a single key more natural (and less error prone as it is easier to spot a case is missing. Since this is largely opinion based I think we need another opinion to decide the way forward.

> I don't know, maybe I'm overlooking sth big in there?

Yes, the comment is inappropriate because perceived speed of developing a review suggestion doesn't say anything about it's sensibility

Reply:
I was asking about the complications in `unnecessarily complicating the (spreadsheet coverage) test`.

Re: the meta issue - sorry, I should've phrased that differently.

Reply:
Looks like the `description` field is unnecessary. It's contents are going to the scaladoc

Reply:
And since the precondition can also go away, so can the whole `MergeabilityTestCase` trait. `Effect[Unit]` being the second element of the tuple is enough

Reply:
I thought you'd say that :) Though to validate that the Name matches the MergeabilityCase we need a common interface to provide the name to - assuming we want to retain alignment and readability. To elaborate - we can either do:

`"!4 !4 -> HadItsMatch(X, Y, Z)` and in the test:
`(mergeabilityCase.checkMatchesName(name) >> mergeabilityCase.assertion).attempt`

or

`hadItsMatch("!4 !4", X, Y, Z)` and in the test:
`value.attempt`

I prefer the former, as the latter reads worse and aligns worse

Reply:
Makes sense, although it's going to need to wait